# **Tugas Praktikum**


Load dataset MNIST dari Keras.

Bangun model dengan 2 hidden layer.

Latih model dan evaluasi akurasi.

In [ ]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import time

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(256, activation='relu'), # Hidden layer 1 dengan 256 neuron
    Dense(128, activation='relu'), # Hidden layer 2 dengan 128 neuron
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
print("\nTraining Model...")
start_time = time.time()
model.fit(X_train, y_train, epochs=10, batch_size=32, verbose=1)
end_time = time.time()
training_time = end_time - start_time
print(f"Training selesai dalam {training_time:.2f} detik")

# 5. Evaluasi model
print("\nEvaluating Model...")
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data uji: {acc:.4f}")


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)



Training Model...
Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 10ms/step - accuracy: 0.8904 - loss: 0.3683
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 8ms/step - accuracy: 0.9733 - loss: 0.0886
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 15s 8ms/step - accuracy: 0.9814 - loss: 0.0576
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 8ms/step - accuracy: 0.9865 - loss: 0.0416
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 8ms/step - accuracy: 0.9899 - loss: 0.0309
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 15s 8ms/step - accuracy: 0.9920 - loss: 0.0239
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 16s 8ms/step - accuracy: 0.9935 - loss: 0.0204
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 16s 9ms/step - accuracy: 0.9938 - loss: 0.0184
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 15s 8ms/step - accuracy: 0.9943 - loss: 0.0174
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 15s 8ms/step - accuracy: 0.9941 - loss: 0.0171
Training selesai dalam 161.50 detik

Evaluating Model...
Akurasi pada data uji: 0.9787



Ubah jumlah neuron di hidden layer (misal: 256 dan 128).

Tambahkan satu hidden layer lagi.

Bandingkan akurasi dan waktu pelatihan.

In [3]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import time

# 1. Load dataset MNIST (sama seperti sebelumnya)
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
print("Membangun Model dengan 3 Hidden Layers (512, 256, 128 neuron)...")
model_modified = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(512, activation='relu'), # Hidden layer 1: 512 neuron (Diubah)
    Dense(256, activation='relu'), # Hidden layer 2: 256 neuron (Diubah)
    Dense(128, activation='relu'), # Hidden layer 3: 128 neuron (Ditambahkan)
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi model
model_modified.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
print("\nTraining Model Modifikasi...")
start_time = time.time()
# Melatih dengan epochs=10, batch_size=32 (sama seperti model asli)
model_modified.fit(X_train, y_train, epochs=10, batch_size=32, verbose=1)
end_time = time.time()
training_time_modified = end_time - start_time
print(f"Training selesai dalam {training_time_modified:.2f} detik")

# 5. Evaluasi model
print("\nEvaluating Model Modifikasi...")
loss_modified, acc_modified = model_modified.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi Model Modifikasi pada data uji: {acc_modified:.4f}")

Membangun Model dengan 3 Hidden Layers (512, 256, 128 neuron)...

Training Model Modifikasi...


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 22s 10ms/step - accuracy: 0.9027 - loss: 0.3261
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9739 - loss: 0.0850
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 11ms/step - accuracy: 0.9803 - loss: 0.0622
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9868 - loss: 0.0430
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9882 - loss: 0.0362
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 11ms/step - accuracy: 0.9918 - loss: 0.0278
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9922 - loss: 0.0233
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 10ms/step - accuracy: 0.9921 - loss: 0.0262
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 23s 12ms/step - accuracy: 0.9933 - loss: 0.0213
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 26s 14ms/step - accuracy: 0.9952 - loss: 0.0164
Training selesai dalam 214.26 detik

Evaluating Model Modifikasi...
Akurasi Model Modifikasi pada d

Model modifikasi dengan tiga hidden layer (512, 256, 128 neuron) mencapai akurasi uji 0.9817 setelah dilatih selama 214.26 detik dan mendapatkan training loss akhir 0.0164.


Eksperimen dengan fungsi aktivasi Sigmoid vs ReLU.

In [4]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import time

# 1. Load dataset MNIST (sama seperti sebelumnya)
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST dengan Aktivasi SIGMOID
print("Membangun Model dengan 3 Hidden Layers (512, 256, 128 neuron) menggunakan SIGMOID...")
model_sigmoid = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(512, activation='sigmoid'), # Diubah: ReLU -> Sigmoid
    Dense(256, activation='sigmoid'), # Diubah: ReLU -> Sigmoid
    Dense(128, activation='sigmoid'), # Diubah: ReLU -> Sigmoid
    Dense(10, activation='softmax') # Output layer (tetap softmax)
])

# 3. Kompilasi model
model_sigmoid.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
print("\nTraining Model dengan SIGMOID...")
start_time = time.time()
# Melatih dengan epochs=10, batch_size=32
model_sigmoid.fit(X_train, y_train, epochs=10, batch_size=32, verbose=1)
end_time = time.time()
training_time_sigmoid = end_time - start_time
print(f"Training selesai dalam {training_time_sigmoid:.2f} detik")

# 5. Evaluasi model
print("\nEvaluating Model dengan SIGMOID...")
loss_sigmoid, acc_sigmoid = model_sigmoid.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi Model Sigmoid pada data uji: {acc_sigmoid:.4f}")

Membangun Model dengan 3 Hidden Layers (512, 256, 128 neuron) menggunakan SIGMOID...

Training Model dengan SIGMOID...


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 23s 11ms/step - accuracy: 0.7744 - loss: 0.7284
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9546 - loss: 0.1535
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9700 - loss: 0.0960
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9810 - loss: 0.0630
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 11ms/step - accuracy: 0.9851 - loss: 0.0492
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9899 - loss: 0.0336
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 42s 12ms/step - accuracy: 0.9906 - loss: 0.0284
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 22s 12ms/step - accuracy: 0.9941 - loss: 0.0182
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9949 - loss: 0.0165
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9964 - loss: 0.0122
Training selesai dalam 233.05 detik

Evaluating Model dengan SIGMOID...
Akurasi Model Sigmoid pada 